ebmlite/core.py

```diff
@@ -811,6 +811,24 @@ def __repr__(self):
         return "<%s %r at 0x%08X>" % (self.__class__.__name__, self.name,
                                       id(self))
 
+    def __enter__(self):
+        """
+        Enter context manager for this document.
+
+        Expects to clean up `self.stream` object; raises an error if stream
+        object needs no such cleanup.
+        """
+        if not hasattr(self.stream, 'close'):
+            raise TypeError(
+                "invalid stream object for context management"
+                " (expected a stream object with a `close` method)"
+            )
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Close this document on exiting context manager."""
+        self.close()
```

Comment:
It might be worth noting in `__init__()` whether the `Document` was instantiated using a filename or a stream. We may not want to close the stream if it were created outside the context.

Reply:
if the user passes in a file object, then the user can simply not use the document object as a context manager. I think it'd be weirder to have a context manager that sometimes doesn't clean up resources.

Reply:
I'm not sure. Are there other classes with contexts that can be instantiated with a stream instead a filename? If there are others, we should do what they do. If not, I think it less weird if the context doesn't affect things outside its scope, so it'd be better to leave the stream open.

Reply:
h5py seems to behave as you suggest (console image below); I'll make a change to do what they do.
![image](https://user-images.githubusercontent.com/5862803/96644928-ca964580-12f7-11eb-9dd4-1754000a1bbe.png)

Reply:
If it were me, I'd implement a flag variable, something like `Document._newfile`, which gets set `True` if the `Document` was instantiated from a filename. That would get checked by the context or by `Document.close()`. The latter would be more like h5py.